### ISSUE TITLE
Visual shader editor crash when any shader node connects back to itself

### ISSUE THREAD

**Issue by RedMser** [godotengine/godot#25098]
**Godot version:**
Latest master - 7ae4cb7f7dc9913f6b1c7e9c9ff90a481ce93ba7

**OS/device including version:**
Windows 7 x64, GeForce GTX 660, OpenGL ES 3.0

**Issue description:**
A node which connects back to itself crashes Godot with a StackOverflowException. Only tested with a single node, but it *most likely* also happens if a chain of nodes loops back to itself (such as `a -> b -> c -> a`)

This can be fixed by detecting a connection which connects back to a parent node (or itself), and highlighting this as an error to the user. In Blender, this is done by changing the color of the connection to red.

The error is found in `VisualShader::_write_node` recursively evaluating its inputs without any security checks.

**Steps to reproduce:**
1. Open project from zip below, BaseScene.tscn
2. Select the mesh -> material -> visual shader -> edit
3. Connect `op` to `b` (ensure that *either* the eye icon is enabled, *or* `op` was connected to `Albedo` before connecting it to `b` - not sure why this causes the shader node's value to get evaluated once...)

**Minimal reproduction project:**
[mrp-shaderInfiniteLoopCrash.zip](https://github.com/godotengine/godot/files/2772985/mrp-shaderInfiniteLoopCrash.zip)

**Comment by Chaosus** [comment]
Looks like a duplicate of #6307 where the same problem is described.

**Comment by RedMser** [comment]
Indeed, I only searched for visual *shaders*, and could not find any reports there.

The difference seems to be that this bug **crashes the editor** the moment you connect the nodes (as it instantly tries to compute a preview value).

The visual *script* bug only causes an error which shows in the debugger when you run the game, and only once it tries computing that node.
